### PR TITLE
Issue #12601: removes code for mutation in EmptyLineSeparator

### PR DESCRIPTION
**Pull request by rnveach**
Issue #12601

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 1]
```diff
@@ -621,14 +621,12 @@ private List<Integer> getEmptyLines(DetailAST ast) {
      */
     private static List<Integer> getEmptyLinesToLog(List<Integer> emptyLines) {
         final List<Integer> emptyLinesToLog = new ArrayList<>();
-        if (emptyLines.size() >= 2) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:624]: I assume this became a surviving mutation with other kills.

This prevents lines with only 1 empty line to be excluded.

The way the code is written, 1 line will never get added to the result. `previousEmptyLineNo` is set to the value of that 1 line. Then the condition to add to the result is `previousEmptyLineNo + 1 == emptyLineNo` which is based on the same 1 line. It logically impossible to for the equality `X + 1 == X` to be true.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 2]
```diff
@@ -621,14 +621,12 @@ private List<Integer> getEmptyLines(DetailAST ast) {
...
-            int previousEmptyLineNo = emptyLines.get(0);
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:625]: This follows the same logic as the previous one.

The for loop still investigates the 1st item and it is impossible to match it as I explained before to add to the result.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 3]
```diff
@@ -619,16 +619,14 @@ private List<Integer> getEmptyLines(DetailAST ast) {
      * @param emptyLines list of empty lines.
      * @return list of empty lines to log.
      */
-    private static List<Integer> getEmptyLinesToLog(List<Integer> emptyLines) {
+    private static List<Integer> getEmptyLinesToLog(Iterable<Integer> emptyLines) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:622]: https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=12734&view=logs&j=eb841e9f-6e1e-5c29-3ad5-74ec2c658069&t=a9b761e6-b8d5-50cf-467a-54967189cab2&l=1262
PMD said this had to change.

**Comment by rnveach** [comment]
Regression coming here after generation.

**Comment by rnveach** [comment]
Regression: http://rveach.no-ip.org/checkstyle/regression/328/
No differences.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 3 above)
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:622]: This link isn't working for me, is it related to removing the call to `size()` and being able to use an interface higher in the inheritance chain?

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java:622]: > is it related to removing the call to size() and get() and being able to use an interface higher in the inheritance chain?

Yes, it is similar to if I use `ArrayList` as a type, but could downgrade it to `List` or even further as `Collection`.

**Review by romani — APPROVED** [review]
ok to merge

**Comment by rnveach** [comment]
@nrmancuso Last one. With approval I will rebase and remove the file completely.

**Review by nrmancuso — APPROVED** [review]

**Pull request merged**